### PR TITLE
ENH: stats: add truncated (i.e. upper bounded) Pareto distribution (truncpar…

### PR DESCRIPTION
**Pull request by tartopohm**
…eto)

#### What does this implement/fix?
This adds a bounded Pareto distribution to the list of continuous distributions, with the class name `truncpareto`.

#### Additional information
* The _fitstart method is overloaded to return saner default.

* No specific unit tests added; the stats module test all pass. I had to add `truncpareto` to the distributions skipped by `test_continuous_basic::check_fit_args_fix` (nan/inf moments), even though the test, when reproduced in a local notebook (scipy 1.8.1-1), passes without triggering any exception (attached file).

2022-05-30T15:30:30

### CODE REVIEW

**Comment by mdhaber** [comment]
This looks like a good start. It looks like this is your first contribution - nice work finding your way around! Do you have suggestions for the guide about adding a new distribution (gh-12133)?

Usually, I'd prefer not to add new truncated distributions because I think the distribution infrastructure should just support truncation parameters (e.g. https://github.com/mdhaber/scipy/pull/75). But it looks like you've even defined `entropy` and moments, so this is more valuable than a generic implementation.

Please send an email to the [mailing list](https://mail.python.org/archives/list/scipy-dev@python.org/) proposing this addition to see if others agree that it's important to add.

**Comment by Tartopohm** [comment]
> 

Hey, thanks for the feedback, I'll open a thread on the list (any guidelines, or free format?).

I needed this distribution for my work, so I scrapped the few necessary methods in a local file, and then it made sense to push everything upstream as soon as I had a bit of time.

The guide was rather straightforward, maybe just a few additional links (things like how to build the doc or how to run the tests/what tests to run) might help. Most of the problems I created were solved by poking the tests around (e.g., using scalar types instead of arrays for the shapes).

I see that some tests are failing (I did not run the slow tests on my computer) should I have a look into it and submit another PR?
Problems seem to be
* fitting (parameters too far away from the reference; but graphically (CDF/ECDF, PDF/histogram), it seems good enough);
* I did not modify doc/source/tutorial/stats.rst to change the number of distributions (oops, I did see it in the instructions, but I skipped it).

**Comment by josef-pkt** [comment]
powerlaw has decreasing pdf for 0 < a < 1

**Comment by josef-pkt** [comment]
I guess they are different, more like the complement, cdf of truncated pareto looks more like and sf of powerlaw.

**Comment by Tartopohm** [comment]
> powerlaw has decreasing pdf for 0 < a < 1

Indeed, that was too quick a look, my bad!

> I guess they are different, more like the complement, cdf of truncated pareto looks more like and sf of powerlaw.

From the doc, SciPy's `powerlaw` is a special case of the Beta, X ~ powerlaw(a) <=> X ~ B(a, 1); while the Pareto is the inverse of that special case, X ~ Pareto(a) <=> 1/X ~ B(a, 1) (hence the standard support starting at 1). This PR truncates the latter.

**Comment by josef-pkt** [comment]
Yes, it looks like truncated pareto is different from powerlaw.
I had forgotten that for pareto the sf is a simple power function and not the cdf.

> X ~ Pareto(a) <=> 1/X ~ B(a, 1)

where did you find this? It's not on the Wikipedia page for Pareto (AFAICS)

**Comment by Tartopohm** [comment]
> Yes, it looks like truncated pareto is different from powerlaw. I had forgotten that for pareto the sf is a simple power function and not the cdf.
> 
> > X ~ Pareto(a) <=> 1/X ~ B(a, 1)
> 
> where did you find this? It's not on the Wikipedia page for Pareto (AFAICS)

I could not find my original source; but from the Wikipedia page for the exponential distribution,
X ~ Exp(a) <=> e^{-X} ~ Beta(a, 1)
and
Y ~ Pareto(a) <=> log(Y) ~ Exp(a).

Putting the two together, unless I am mistaken, Z ~ Pareto(a) <=> log(Z) ~ Exp(a) <=> -log(1/Z) ~ Exp(a) <=> 1/Z ~ Beta(a, 1).

File: scipy/stats/_continuous_distns.py
Comment (mdhaber) [inline scipy/stats/_continuous_distns.py:8614]: This reference doesn't appear to have the formulas for all of the overridden methods below. Did you derive those yourself? If so, could you provide your notes? If not, could you point to a specific reference with each? As a reviewer, I would check all of the methods against their source, so whatever you can do to facilitate that process would be appreciated.

File: scipy/stats/_continuous_distns.py
Comment (mdhaber) [inline scipy/stats/_continuous_distns.py:8666]: This is fine, and there's no need to check if you're not interested, but I'm curious whether the notes [here](https://github.com/scipy/scipy/issues/12545#issuecomment-735313963) can be extended to derive analytical fit including the truncation parameter?

Comment (Tartopohm) [reply scipy/stats/_continuous_distns.py:8666]: The Lagrange multipliers approach is pretty cool!
I had a quick look; it seems that the constraints on the domain reduce to (I use s for scale and tau for location)
image,
which gives
image.
We are left with an optimization problem on b and tau.

I will need to look into it in more details and double-check my derivations.

File: scipy/stats/_continuous_distns.py
Comment (Tartopohm) [reply scipy/stats/_continuous_distns.py:8614]: To be honest, I mostly put this reference for context/example of application.
I did most of the derivations; it was easier to accommodate SciPy's framework this way.
It was a bit of a scribbly mess, so I took the time to clean it up! All the overridden methods, except the _log[pdf|cdf|sf] (really straightforward to derive) should be covered in the attached file :)

[notes.pdf](https://github.com/scipy/scipy/files/8832762/notes.pdf)

File: scipy/stats/_continuous_distns.py
Comment (mdhaber) [inline scipy/stats/_continuous_distns.py:8622]: Mostly good - not integral, neither end of the domain is inclusive, and b is positive, but c has to be greater than 1.
```suggestion
        ic = _ShapeInfo("c", False, (1.0, np.inf), (False, False))
```

File: scipy/stats/_continuous_distns.py
Comment (mdhaber) [inline scipy/stats/_continuous_distns.py:8632]: 👍

**Review by mdhaber — COMMENTED** [review]
Didn't make it very far - but I started!

<details>

```python3
from sympy import symbols, integrate, simplify
from IPython.display import display
c, x = symbols('c, x', real=True)
b = symbols('b', real=True, positive=True)


pdf = b * x**-(b+1) 
cdf = integrate(pdf, (x, 1, c))

pdft = pdf / cdf
cdft = simplify(integrate(pdft, (x, 1, c)))
```

</details>

File: scipy/stats/_continuous_distns.py
Comment (mdhaber) [inline scipy/stats/_continuous_distns.py:8641]: 👍

File: scipy/stats/_continuous_distns.py
Comment (mdhaber) [inline scipy/stats/_continuous_distns.py:8670]:
```suggestion
        b, loc, scale = pareto.fit(data)
        c = (np.max(data) - loc)/scale
```
I think this would be a little simpler.

File: scipy/stats/_distr_params.py
Comment (mdhaber) [inline scipy/stats/_distr_params.py:108]: These look like reasonable parameters.
image

File: scipy/stats/_distr_params.py
Comment (mdhaber) [inline scipy/stats/_distr_params.py:267]: 👍

File: scipy/stats/tests/test_fit.py
Comment (mdhaber) [inline scipy/stats/tests/test_fit.py:444]: 👍

**Review by mdhaber — COMMENTED** [review]
Still looking good so far. I need to check the remaining methods, the end of `_fitstart`, the tutorial, and the checklist (to make sure I'm not forgetting something).

File: scipy/stats/_continuous_distns.py
Comment (mdhaber) [inline scipy/stats/_continuous_distns.py:8657]: 👍

File: scipy/stats/_continuous_distns.py
Comment (mdhaber) [inline scipy/stats/_continuous_distns.py:8663]: 👍

**Review by mdhaber — COMMENTED** [review]
Finished checking the methods using Sympy. Sympy has trouble verifying the `log` functions, but they are clearly correct.

<details>

```python3
from sympy import symbols, integrate, simplify, solve, log
from IPython.display import display
b, c, x, q = symbols('b, c, x, q', real=True, positive=True)
n = symbols('n', real=True, positive=True, integer=True)

def log1p(f):
    return log(1 + f)


pdfp = b * x**-(b+1) 
cdfp = integrate(pdfp, (x, 1, c))

pdf = pdfp / cdfp
pdf2 = b * x**-(b+1) / (1 - c**-b)
assert simplify(pdf - pdf2) == 0

logpdf = log(pdf)
logpdf2 = log(b) - log1p(-c**-b) - (b+1)*log(x)
display(simplify(logpdf - logpdf2))  # obviously equal 
assert logpdf.equals(logpdf2) is None

cdf = simplify(integrate(pdf, (x, 1, x)))
cdf2 = (1 - x**-b) / (1 - c**-b)
assert simplify(cdf - cdf2) == 0

logcdf = log(cdf)
logcdf2 = log1p(-x**-b) - log1p(-c**-b)
display(simplify(logcdf - logcdf2))  # obviously equal 
# assert logcdf.equals(logcdf2) is None

ppf = solve(cdf - q, x)[0]
ppf2 = (1 - (1 - c**-b)*q)**(-1/b)
assert simplify(ppf2.subs(q, cdf)) - x == 0
# assert simplify(ppf - ppf2) == 0

sf = 1 - cdf
sf2 = (x**-b - c**-b) / (1 - c**-b)
assert simplify(sf2 - sf) == 0

logsf = log(sf)
logsf2 = log(x**-b - c**-b) - log1p(-c**-b)
display(simplify(logsf - logsf2))  # obviously equal
assert logsf.equals(logsf2) is None

isf = solve(sf - q, x)[0]
isf2 = (c**-b + (1 - c**-b)*q)**(-1/b)
assert simplify(isf2.subs(q, sf)) - x == 0

ent = -integrate(pdf * log(pdf), (x, 1, c))
ent2 = -(log(b/(1 - c**-b)) + (b+1)*(log(c)/(c**b - 1) - 1/b))
assert simplify(ent - ent2) == 0

munp = integrate(pdf * x**n, (x, 1, c))
munpA = simplify(munp.args[0][0])
munpB = simplify(munp.args[1][0])
munp2A = b / (b-n) * (c**b - c**n) / (c**b - 1)
munp2B = b*log(c) / (1 - c**-b)
assert simplify(munpA - munp2A) == 0
assert simplify(munpB - munp2B) == 0
```


</details>

File: scipy/stats/_continuous_distns.py
Comment (mdhaber) [inline scipy/stats/_continuous_distns.py:8674]: I would just
```suggestion
        return b, c, loc, scale
```
and not even acccept `args` into this override of `fitstart`. Some overrides of `_fitstart` accept `args`, but many do not,, and that's OK because it's never used.
(`rv_continuous._fitstart` accepts `args` to allow the possibility of overrides to give guesses for the shape parameters but leave it to `rv_continuous._fitstart` to determine guesses for the `loc` and `scale`, but overrides of `_fitstart` don't need to.)

**Comment by mdhaber** [comment]
All boxes in the checklist seem to be satisfied.

File: scipy/stats/_continuous_distns.py
Comment (mdhaber) [reply scipy/stats/_continuous_distns.py:8606]: Did you not like the way that `truncnorm` explains the truncation definition? You wrote this differently:
image

and it's clear, but there is always the issue of how to refer to `loc` and `scale` in math. Italicized words "loc" and "scale" are not quite right. You could un-italicize them, or even teletype them, in the math, or you could define single-letter variables to represent them in the math.

File: doc/source/tutorial/stats/continuous_truncpareto.rst
Comment (mdhaber) [inline doc/source/tutorial/stats/continuous_truncpareto.rst:26]: Missing a negative sign.

**Review by mdhaber — CHANGES_REQUESTED** [review]
I think this will be ready to merge after the comments are addressed.

File: scipy/stats/_continuous_distns.py
Comment (mdhaber) [reply scipy/stats/_continuous_distns.py:8632]: There might be numerical advantages to evaluating this as the exponentiated logpdf, but that could be left as an enhancement if we're unsure.

File: scipy/stats/_continuous_distns.py
Comment (Tartopohm) [reply scipy/stats/_continuous_distns.py:8666]: > This is fine, and there's no need to check if you're not interested, but I'm curious whether the notes [here](https://github.com/scipy/scipy/issues/12545#issuecomment-735313963) can be extended to derive analytical fit including the truncation parameter.

There are definitely optimisations to be made regarding fitting. I have yet to check for combinations involving more than one fixed parameter, but for none or one provided, `root_scalar` should do the trick.

I'll address the requested changes and commit along the improved `fit` whenever I can.

**Comment by mdhaber** [comment]
Please don't force push after a review begins unless a reviewer has agreed that you should. (Hopefully I have a copy of your branch so I can see what changed since my last review, otherwise I'd have to start over.) Thanks.

**Comment by Tartopohm** [comment]
> Please



> Please don't force push after a review begins unless a reviewer has agreed that you should. (Hopefully I have a copy of your branch so I can see what changed since my last review, otherwise I'd have to start over.) Thanks.

Apologies, I am really not a git wiz, still missing a lot on the subtlety

File: scipy/stats/_continuous_distns.py
Comment (mdhaber) [inline scipy/stats/_continuous_distns.py:8636]: Here's how this is rendering:
image

According to the numpydoc style guide, single backticks are appropriate for "variables". If we want code to render in monospaced font, though, double backticks are needed. 

These rules are really ambiguous, but here is how I normally would apply them.
```suggestion
    Notice that the upper truncation value :math:`c` is defined in
    standardized form so that random values of an unscaled, unshifted variable
    are within the range ``[1, c]``.
    If ``u_r`` is the upper bound to a scaled and/or shifted variable,
    then ``c = (u_r - loc) / scale``. In other words, the support of the
    distribution becomes ``(scale + loc) <= x <= (c*scale + loc)`` when
    `scale` and/or `loc` are provided.
```

As parameters of the distribution, `scale` and `loc` are definitely "variables", so I've left the single backticks alone. (Why they shouldn't always render in monospaced font, and why it is correct for them to render in bold
image
italics,
image
or monospaced font
image
depending on context is beyond me, so don't worry that this was not intuitive.)

Because `c = (u_r - loc) / scale` and `(scale + loc) <= x <= (c*scale + loc)` should render in monospaced font, I've given them double backticks.

`u_r` is questionable. I suppose it's a "variable"? But it's not actually a parameter or even a variable in the code, so since I would prefer it to render in monospaced font as it does in `c = (u_r - loc) / scale`, I've given it double backticks. That's a judgment call.

**Review by mdhaber — COMMENTED** [review]
(CI failures are unrelated.)

It worked out - I had a local copy and was able to verify that all the changes were made appropriately. I'm going to commit this minor correction to the docstring and wait for docs to build. If they look good, shall I merge? (All set on your end? Did you agree with the previous round of comments?)

**Review by mdhaber — APPROVED** [review]
I think this is all set. Let me know if you're happy with it @Tartopohm and I'll merge.

**Comment by Tartopohm** [comment]
> I think this is all set. Let me know if you're happy with it @Tartopohm and I'll merge.

All good for me! Thanks

**Comment by mdhaber** [comment]
Thanks @Tartopohm! 
Would you be interested in contributing any other univariate distributions? There are lots of open enhancement requests, e.g. 
gh-10364
gh-13825
gh-14806
gh-11291
gh-6414

**Pull request merged**
